src/proxy/connection_manager.rs

```diff
@@ -54,24 +56,37 @@ impl ConnectionManager {
         }
     }
 
+    // register a connection with the manager and get a channel to receive on
     pub async fn track(self, c: &Connection) -> drain::Watch {
-        // consider removing this whole if let since if it's None we need to perform another get inside the write lock to prevnt racy inserts
-        if let Some(cd) = self.drains.read().await.get(c) {
-            return cd.rx.clone();
-        }
         let cd = ConnectionDrain::new();
         let rx = cd.rx.clone();
-        //TODO: this was racy, another insert may happen between dropping the read lock and attaining this write lock
-        // try_insert is the best solution once it's no longer a nightly-only experimental API
         let mut drains = self.drains.write().await;
-        if let Some(w) = drains.get(c) {
-            return w.rx.clone();
+        if let Some(w) = drains.remove(c) {
+            w.count.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+            let rx = w.rx.clone();
+            drains.insert(c.clone(), w);
+            debug!("there are {} tracked connections", drains.len());
+            return rx;
         }
         drains.insert(c.clone(), cd);
+        debug!("there are {} tracked connections", drains.len());
         rx
     }
 
-    pub async fn close(&self, c: &Connection) {
+    // releases tracking on a connection
+    // uses a counter to determine if there are other tracked connections or not so it may retain the tx/rx channels when necessary
+    pub async fn release(self, c: &Connection) {
+        let mut drains = self.drains.write().await;
+        if let Some((k, v)) = drains.remove_entry(c) {
+            if v.count.fetch_sub(1, std::sync::atomic::Ordering::SeqCst) > 1 {
+                // something else is tracking this connection, retain
+                drains.insert(k, v);
```

Comment:
We talked about this already - but `drain-rs` is a _very_ light and rather inflexible wrapper around plain old tokio channels, and I suspect there might be other ways to accomplish this by using tokio channels directly (different channel type, or something like `WeakSender`), without the manual external atomic count.

But the effect would be the same so it is probably not worth it, this will do the job AFAICT - but if `drain-rs` gives you trouble and it's easier to drop, I don't have a problem with ditching it.

Reply:
It's a fair idea. I'm not sure we strictly need the bidirectional close functionality the crate offers for this case... in fact we may prefer a fire and forget since for our purposes right now we don't actually care if the receiver get's a signal or an error. We're sending 1 bit of information which just means do your thing... At any rate this is certainly something to consider.